lib/Conversion/RTLToLLHD/RTLToLLHD.cpp

```diff
@@ -0,0 +1,178 @@
+  target.addLegalDialect<LLHDDialect>();
+  target.addIllegalOp<RTLModuleOp>();
+
+  RTLToLLHDTypeConverter typeConverter;
+  OwningRewritePatternList patterns;
+  patterns.insert<ConvertRTLModule>(typeConverter, &context);
+  patterns.insert<ConvertOutput>(typeConverter, &context);
+
+  if (failed(applyPartialConversion(module, target, std::move(patterns))))
+    signalPassFailure();
+}
+
+//===----------------------------------------------------------------------===//
+// TypeConverter conversions and materializations
+//===----------------------------------------------------------------------===//
+RTLToLLHDTypeConverter::RTLToLLHDTypeConverter() {
+  // Convert IntegerType by just wrapping it in SigType.
+  addConversion([](IntegerType type) { return SigType::get(type); });
+}
+
+//===----------------------------------------------------------------------===//
+// Convert structure operations
+//===----------------------------------------------------------------------===//
+
+/// This works on each RTL module, creates corresponding entities, moves the
+/// bodies of the modules into the entities, and converts the bodies.
+struct ConvertRTLModule : public OpConversionPattern<RTLModuleOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(RTLModuleOp module, ArrayRef<Value> operands,
+                                ConversionPatternRewriter &rewriter) const {
+    // Collect the RTL module's port types in a signature conversion.
+    FunctionType moduleType = module.getType();
+    unsigned numInputs = moduleType.getNumInputs();
+    TypeConverter::SignatureConversion argConversion(module.getNumArguments() +
+                                                     module.getNumResults());
+
+    if (failed(typeConverter->convertSignatureArgs(moduleType.getInputs(),
+                                                   argConversion)) ||
+        failed(typeConverter->convertSignatureArgs(moduleType.getResults(),
+                                                   argConversion, numInputs)))
+      return module.emitError("type converter failed to convert signature");
+
+    // Create the entity.
+    auto entity = rewriter.create<EntityOp>(module.getLoc(), numInputs);
+
+    // Inline the RTL module body into the entity body.
+    Region &entityBodyRegion = entity.getBodyRegion();
+    rewriter.inlineRegionBefore(module.getBodyRegion(), entityBodyRegion,
+                                entityBodyRegion.end());
+
+    // Attempt to convert the entry block's args using the previous conversion.
+    if (failed(rewriter.convertRegionTypes(&entityBodyRegion, *typeConverter,
+                                           &argConversion)))
+      return module.emitError("could not convert region types");
+
+    // Add the LLHD terminator op after the RTL module's output ops.
+    rewriter.setInsertionPointToEnd(entity.getBodyBlock());
+    rewriter.create<llhd::TerminatorOp>(entity.getLoc());
```

Comment:
This is super nice. I really like all the infrastructure that MLIR keeps around for this kind of task. Very neat 😄